### PR TITLE
E2-10023 add trackGeofence

### PR DESCRIPTION
**Pull request by gg4race**
https://leanplum.atlassian.net/secure/RapidBoard.jspa?rapidView=28&projectKey=E2&modal=detail&selectedIssue=LP-10023

### CODE REVIEW

File: Leanplum-SDK/Classes/Internal/Leanplum.m [hunk 1]
```diff
@@ -1948,7 +1935,25 @@ + (void)track:(NSString *)event withValue:(double)value andInfo:(NSString *)info
...
+}
 
++ (void)track:(NSString *)event withValue:(double)value andInfo:(NSString *)info
+      andArgs:(NSDictionary *)args andParameters:(NSDictionary *)params
+{
+    RETURN_IF_NOOP;
+    LP_TRY
+    
+    // Track should not be called in background.
+    if (![NSThread isMainThread]) {
```
Comment (hrishileanplum) [inline Leanplum-SDK/Classes/Internal/Leanplum.m]: Interesting. I know this is existing restriction.

File: Leanplum-SDK/Classes/Internal/Leanplum.m [hunk 2]
```diff
@@ -1948,7 +1935,25 @@ + (void)track:(NSString *)event withValue:(double)value andInfo:(NSString *)info
...
+        dispatch_sync(dispatch_get_main_queue(), ^{
+            [self track:event withValue:value andInfo:info andArgs:args andParameters:params];
+        });
+        return;
```
Comment (hrishileanplum) [inline Leanplum-SDK/Classes/Internal/Leanplum.m]: The problem is , we are not throwing any errors stating that "you have called track on main thread etc", it just returns. I know this is existing code, we might have to think on this together, cos of the way the callbacks are setup.

File: Leanplum-SDK/Classes/Internal/Leanplum.m [hunk 3]
```diff
@@ -1957,6 +1962,37 @@ + (void)track:(NSString *)event withValue:(double)value andInfo:(NSString *)info
...
+    } else {
+        [[LPCountAggregator sharedAggregator] incrementCount:@"track_geofence_disabled"];
+    }
+}
+
++ (void)trackGeofence:(NSString *)event withValue:(double)value andInfo:(NSString *)info
+                           andArgs:(NSDictionary *)args andParameters:(NSDictionary *)params
+{
+    RETURN_IF_NOOP;
+    LP_TRY
```
Comment (hrishileanplum) [inline Leanplum-SDK/Classes/Internal/Leanplum.m]: can you help me understand the reason for these 2 constants or macros?

File: Example/Tests/Classes/LeanplumTest.m [hunk 4]
```diff
@@ -49,6 +49,8 @@ @interface Leanplum (Test)
 + (NSSet<NSString *> *)parseEnabledFeatureFlagsFromResponse:(NSDictionary *)response;
 + (void)triggerMessageDisplayed:(LPActionContext *)context;
 
++ (void)trackGeofence:(NSString *)event withValue:(double)value andInfo:(NSString *)info andArgs:(NSDictionary *)args andParameters:(NSDictionary *)params;
```
[the remark this answers is not in the document]
Comment (hrishileanplum) [reply Example/Tests/Classes/LeanplumTest.m]: Sample example.  string representation logic for enums.
https://stackoverflow.com/questions/1094984/convert-objective-c-typedef-to-its-string-equivalent

File: Leanplum-SDK/Classes/Internal/Leanplum.m [hunk 5]
```diff
@@ -1993,6 +1995,23 @@ + (void)trackGeofence:(NSString *)event withValue:(double)value andInfo:(NSStrin
     LP_END_TRY
 }
 
++ (NSString *)getEventNameFromGeofenceType:(LPGeofenceEventType)event {
```
Comment (hrishileanplum) [inline Leanplum-SDK/Classes/Internal/Leanplum.m]: Just 1 thing.
Can we please move the entire Enum logic  to a separate file and expose an Api for getting the name similar to Android. You can place the file in 
"Internal" directory.
 Something like this.

/  EnumConstants.h


#ifndef EnumConstants_h
#define EnumConstants_h

typedef enum {
    LPEnterRegion,
    LPExitRegion
} LPGeofenceEventType;
#endif

File: Leanplum-SDK/Classes/Internal/Leanplum.m (hunk 3 above)
Comment (gg4race) [reply Leanplum-SDK/Classes/Internal/Leanplum.m]: I copied them from the `track` method. I think they were probably added after experiencing issues/crashes, not sure of the exact reason.

**Comment by gg4race** [comment]
same as android - We should call sendIfConnected instead of send because sendIfConnected will try to send to the server right away whereas send will only save request to database. (See design doc, we want to alert server as soon as geofence is crossed)

**Pull request merged**